crates/backend/src/codegen.rs

```diff
@@ -1024,114 +1024,113 @@ impl ToTokens for ast::ImportType {
 impl ToTokens for ast::ImportEnum {
     fn to_tokens(&self, tokens: &mut TokenStream) {
         let vis = &self.vis;
-        let name = &self.name;
-        let expect_string = format!("attempted to convert invalid {} into JSValue", name);
+        let enum_name = &self.name;
+        let name_str = enum_name.to_string();
+        let name_len = name_str.len() as u32;
+        let name_chars = name_str.chars().map(|c| c as u32);
         let variants = &self.variants;
-        let variant_strings = &self.variant_values;
+        let variant_count = self.variant_values.len() as u32;
+        let variant_values = &self.variant_values;
+        let variant_indices = (0..variant_count).collect::<Vec<_>>();
+        let invalid = variant_count;
+        let hole = variant_count + 1;
         let attrs = &self.rust_attrs;
 
-        let mut current_idx: usize = 0;
-        let variant_indexes: Vec<Literal> = variants
-            .iter()
-            .map(|_| {
-                let this_index = current_idx;
-                current_idx += 1;
-                Literal::usize_unsuffixed(this_index)
-            })
-            .collect();
-
-        // Borrow variant_indexes because we need to use it multiple times inside the quote! macro
-        let variant_indexes_ref = &variant_indexes;
-
         // A vector of EnumName::VariantName tokens for this enum
         let variant_paths: Vec<TokenStream> = self
             .variants
             .iter()
-            .map(|v| quote!(#name::#v).into_token_stream())
+            .map(|v| quote!(#enum_name::#v).into_token_stream())
             .collect();
 
         // Borrow variant_paths because we need to use it multiple times inside the quote! macro
         let variant_paths_ref = &variant_paths;
 
         let wasm_bindgen = &self.wasm_bindgen;
 
+        let describe_variants = self.variant_values.iter().map(|variant_value| {
+            let length = variant_value.len() as u32;
+            let chars = variant_value.chars().map(|c| c as u32);
+            quote! {
+                inform(#length);
+                #(inform(#chars);)*
+            }
+        });
+
         (quote! {
             #(#attrs)*
-            #vis enum #name {
-                #(#variants = #variant_indexes_ref,)*
+            #[non_exhaustive]
```

Comment:
Should we explore removing the non_exhaustive here?

Reply:
What was your motivation to add it in the first place?

Reply:
as per https://rust-lang.github.io/rfcs/2008-non-exhaustive.html, __Nonexhaustive was the old way of implementing non-exhaustive enums so I wanted to preserve the original behavior to limit the scope of this PR.

Reply:
Don't get me wrong, I think using `#[non_exhaustive]` is the right thing to do, but we didn't remove the hidden variant, which we can't for backwards compatibility reasons, so it seems to be unrelated to this PR.

Or am I missing something here?
Either way I'm fine with leaving it there.

Reply:
I'm realizing that nothing is currently stopping users from matching the __Nonexhaustive variant. In the link I shared, it seems like the unstable attribute is what's needed to make it unmatchable. So apparently adding non_exhaustive here is a breaking change.

Reply:
We did mark it as `#[doc(hidden)]` and its name starts with `__`, this intents to clearly mark it as unstable. This is unfortunately what has been done in the past before `#[non_exhaustive]` existed, so I think we are in the clear here, even though this is not clearly specified anywhere.

Unfortunately we aren't able to use the `#[unstable]` attribute, so it is what it is.
I think ideally here we would neither need `#[non_exhaustive]` nor the extra variant and make it so it panics/throws instead of constructing this variant. Unfortunately this would not be backwards compatible, so we are stuck with this until the next breaking change.

Reply:
In the case of WebIDL, yes.
But in the case of user defined string enums it should be possible to have them be exhaustive.

Reply:
Ok yeah, so the webidl->rust generator would add the attribute instead